Adds charts in "charts\" directory to dependencies

sushilkm opened the pull request:
Dependencies should be processed even if they are not added to requirements.yaml
Fixes https://github.com/kubernetes/helm/issues/2599

Commit:
Adds charts in "charts\" directory to dependencies

Dependencies should be processed even if they are not added to requirements.yaml
Fixes https://github.com/kubernetes/helm/issues/2599

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/helm/2607)
<!-- Reviewable:end -->

jascott1 commented on pkg/chartutil/requirements.go:
```diff
@@ -258,6 +258,24 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
+	// then this should be added to chartDependencies.
+	// However, if the dependency is already specified in requirements.yaml
+	// we should not add it, as it would be anyways processed from requirements.yaml
+
```
seems verbose, just my imo but consider shortening to something like
"// add dependencies from /charts not already present in requirements.yaml"

technosophos replied:
I kinda prefer the more verbose description, since I have the long-term memory of a sea slug. :-) But I'm okay with whatever.

technosophos replied:
ROTFL. After writing that, I googled "Sea Slug Long Term Memory"... and discovered that they actually do memory research on sea slugs. https://www.livescience.com/17683-sea-slug-memory.html

jascott1 replied:
Part of the new code review process? "They tested the model by submitting the animals to a series of training sessions, involving electric shocks"

technosophos commented:
We should catalog some test cases we need to verify. Off the top of my head, it's:

- subcharts, no requirements.yaml
- subcharts, all listed in requirements.yaml
- subcharts, some in requirements, some not.

Am I missing and edge cases or other pertinent scenarios? Do we need to do testing based on whether or not charts are enabled/disabled by a particular requirements file?

chancez commented:
Might be good to validate that chart's containing subcharts, that have subcharts, too. Not something I use, but some limited testing of various nesting cases is probably good.

Commit:
Added tests for different combinations of subcharts & requirements.yaml

Added more tests to test following comibnations of subcharts and requirements.yaml
- [ ] subcharts not specified in requirements.yaml
- [ ] some subcharts specified in requirements.yaml while others should be picked directly from "charts\"
- [ ] all subcharts specified in requirements.yaml without alias
- [ ] subcharts specified in requirements.yaml with alias is already tested

jascott1 commented:
@sushilkm You might want to checkout the testdata/subpop chart as it already has a bunch of nested dependencies. You could add some charts that aren't in a requirements.yaml.

sushilkm commented:
@jascott1 I think the only case `testdata/subpop` can help is for trying all subcharts specified in `requirements.yaml`, or is there anything else you want me to test with `subpop`

jascott1 commented:
@sushilkm nothing specific, just thought it might be useful to modify and test with since it already has deep nested charts. Sounds like you have it covered :)

thomastaylor312 commented:
Marked this as small since most of the code is just test charts 😄

technosophos commented:
Are we ready to re-review and merge this?

I can break out the low-voltage system and shock some sea slugs @jascott1 ;-)

scottrigby commented:
@sushilkm Great work. Also love the lyrical test data. :shipit:

jascott1 commented:
@technosophos I think it needs re-review now as it seems ready. Thats correct @sushilkm?

sushilkm commented:
Yup waiting for reviews on this one

technosophos commented:
@thomastaylor312 When you marked it "in process", were you claiming this one? If not, I can do a re-review.

thomastaylor312 commented:
@technosophos Sorry no, I was just fixing the label because people were already reviewing it.

technosophos commented:
Okay. Gonna try to get this tested and committed before the k8s 1.7 PR is merged.

technosophos reviewed (approved):
pulled and manually ran tests.
